app/api/server/v1/im.js

```diff
@@ -18,7 +19,7 @@ function findDirectMessageRoom(params, user) {
 		nameOrId: params.username || params.roomId,
 	});
 
-	const canAccess = Meteor.call('canAccessRoom', room._id, user._id);
+	const canAccess = Meteor.call('canAccessRoom', room._id, user._id) || hasPermission(user._id, 'view-room-administration');
```

Comment:
If i'm following to code flow from server/methods/canAccessRoom.js and server/services/authorization/canAccessRoom.ts correctly this would not allow a user to find a (DM) room that they don't belong to. The main purpose behind this change is to allow for a user (through the API) to manage rooms and room memberships even if they themselves are not in that room. Another solution would be to modify either of the fore-mentioned files to include this permission check there, however that seemed a much more invasive change so I was trying to leave less of a footprint.

Reply:
the problem I see is that this will allow an admin to call any endpoint since he will have the permission to it and this is not what we want.

we have few endpoints that admins are allowed to call that already have this check, for example `dm.messages.others` , `dm.list.everyone`, etc..

Reply:
what do you think of extending the `findDirectMessageRoom` function definition to include a `allowAdmin` parameter (only set for the `im.delete` endpoint) ? Then `canAccess` would become: 
```
const canAccess = Meteor.call('canAccessRoom', room._id, user._id) ||
   (allowAdmin && hasPermission(user._id, 'view-room-administration'));
```

Reply:
looks good..

there might be a missing test then.. because when I was testing this I removed the additional `hasPermission` and all tests were passing correctly, even the `im.delete` ones.

Reply:
That is expected. I put a guard at the top of the `im.delete` route to also check for `view-room-administration` permissions. See https://github.com/RocketChat/Rocket.Chat/pull/18022/files#diff-662ace3def7f08842c41f0735e25ff4eb86e618381d6fcf109e47280fdaba7c6R56-R58

For the associated non-admin test see https://github.com/RocketChat/Rocket.Chat/pull/18022/files#diff-b02596576d756837f387d49b0e36882e5aabf572e653d2839e8f9e20b3182a41R579-R621

I know this means we're now checking this permission twice, but it feels like that is safer given the `findDirectMessageRoom` is used so heavily elsewhere. 

Another option which would remove the need for the "allow admin override" parameter would be to just call the `getDirectMessageByNameOrIdWithOptionToJoin` method in the delete route directly? ie:
```
API.v1.addRoute(['dm.delete', 'im.delete'], { authRequired: true }, {
	post() {
		if (!hasPermission(this.userId, 'view-room-administration')) {
			return API.v1.unauthorized();
		}

		const params = this.requestParams();
		const room = getDirectMessageByNameOrIdWithOptionToJoin({
			currentUserId: this.user._id,
			nameOrId: params.username || params.roomId,
		});
		if (!room || room.t !== 'd') {
			throw new Meteor.Error('error-room-not-found', 'The required "roomId" or "username" param provided does not match any direct message');
		}

		Meteor.call('eraseRoom', room._id);

		return API.v1.success();
	},
});
```
We don't nee any extra permission checks given the view admin guard at the top. There is a little more double-up in loading the room in that route, but likely less confusion overall.

Reply:
oh, I just realised it calls `getDirectMessageByNameOrIdWithOptionToJoin`, which may create the room, before you removes it :thinking: not sure this can happen as it depends on the accepted params.. but by just thinking this can happen is weird.

I don't think you need to do this change, let's keep the double permission check, it is only for this particular endpoint, so I think we're good